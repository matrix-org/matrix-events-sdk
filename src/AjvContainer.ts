/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Ajv, {Schema, SchemaObject} from "ajv";
import AjvErrors from "ajv-errors";
import {NamespacedValue} from "./NamespacedValue";

export class AjvContainer {
    public static readonly ajv = new Ajv({
        allErrors: true,
    });

    static {
        AjvErrors(AjvContainer.ajv);
    }

    /* istanbul ignore next */
    // noinspection JSUnusedLocalSymbols
    private constructor() {}

    /**
     * Creates a JSON Schema representation of the EitherAnd<> TypeScript type.
     * @param ns The namespace to use in the EitherAnd<> type.
     * @param schema The schema to use as a value type for the namespace options.
     * @returns The EitherAnd<> type as a JSON Schema.
     */
    public static eitherAnd<S extends string = string, U extends string = string>(
        ns: NamespacedValue<S, U>,
        schema: Schema,
    ): {anyOf: SchemaObject[]; errorMessage: string} {
        // Dev note: ajv currently doesn't have a useful type for this stuff, but ideally it'd be smart enough to
        // have an "anyOf" type we can return.
        // Also note that we don't use oneOf: we manually construct it through a Type A, or Type B, or Type A+B list.
        if (!ns.altName) {
            throw new Error("Cannot create an EitherAnd<> JSON schema type without both stable and unstable values");
        }
        return {
            errorMessage: `schema does not apply to ${ns.stable} or ${ns.unstable}`,
            anyOf: [
                {
                    type: "object",
                    properties: {
                        [ns.name]: schema,
                    },
                    required: [ns.name],
                    errorMessage: {
                        properties: {
                            [ns.name]: `${ns.name} is required`,
                        },
                    },
                },
                {
                    type: "object",
                    properties: {
                        [ns.altName]: schema,
                    },
                    required: [ns.altName],
                    errorMessage: {
                        properties: {
                            [ns.altName]: `${ns.altName} is required`,
                        },
                    },
                },
                {
                    type: "object",
                    properties: {
                        [ns.name]: schema,
                        [ns.altName]: schema,
                    },
                    required: [ns.name, ns.altName],
                    errorMessage: {
                        properties: {
                            [ns.name]: `${ns.name} is required`,
                            [ns.altName]: `${ns.altName} is required`,
                        },
                    },
                },
            ],
        };
    }
}
