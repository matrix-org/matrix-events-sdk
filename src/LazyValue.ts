/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Represents a lazily-loaded value. When accessed for the first time, the
 * value will be cached for the lifetime of the object. The getter function
 * will be released when the value is cached.
 */
export class LazyValue<T> {
    private cached: T | undefined;
    private getter: (() => T) | undefined;

    public constructor(getter: () => T) {
        // manually copy, so we don't expose the `undefined` type option
        this.getter = getter;
    }

    public get value(): T {
        if (this.getter !== undefined) {
            this.cached = this.getter();
            this.getter = undefined;
        }
        // Force a cast rather than assert because it's possible for `T` to
        // be undefined or null (we don't stop people from making bad decisions
        // in this class).
        return this.cached as T;
    }
}
