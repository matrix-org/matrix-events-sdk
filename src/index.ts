// created from ctix

// created from ctix

export * from "./content_blocks/m/EmoteBlock";
export * from "./content_blocks/m/MarkupBlock";
export * from "./content_blocks/m/NoticeBlock";
export * from "./events/m/EmoteEvent";
export * from "./events/m/MessageEvent";
export * from "./events/m/NoticeEvent";
export * from "./content_blocks/ArrayBlock";
export * from "./content_blocks/BaseBlock";
export * from "./content_blocks/BooleanBlock";
export * from "./content_blocks/IntegerBlock";
export * from "./content_blocks/InvalidBlockError";
export * from "./content_blocks/ObjectBlock";
export * from "./content_blocks/StringBlock";
export * from "./events/InvalidEventError";
export * from "./events/RoomEvent";
export * from "./LazyValue";
export * from "./NamespacedValue";
export * from "./types";
