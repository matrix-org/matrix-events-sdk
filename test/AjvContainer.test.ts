/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {AjvContainer} from "../src/AjvContainer";
import {NamespacedValue} from "../src";

describe("AjvContainer", () => {
    describe("eitherAnd", () => {
        it("should reject a namespaced value missing an altName", () => {
            expect(() => {
                AjvContainer.eitherAnd(new NamespacedValue("stable", null), {type: "object"});
            }).toThrow(
                new Error("Cannot create an EitherAnd<> JSON schema type without both stable and unstable values"),
            );
        });

        it("should generate an appropriate schema type", () => {
            const result = AjvContainer.eitherAnd(new NamespacedValue("stable", "unstable"), {
                type: "object",
                properties: {test: {type: "string"}},
            });
            expect(result).toMatchSnapshot();
        });
    });
});
