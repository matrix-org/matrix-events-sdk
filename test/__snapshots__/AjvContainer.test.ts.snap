// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AjvContainer eitherAnd should generate an appropriate schema type 1`] = `
{
  "anyOf": [
    {
      "errorMessage": {
        "properties": {
          "stable": "stable is required",
        },
      },
      "properties": {
        "stable": {
          "properties": {
            "test": {
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "required": [
        "stable",
      ],
      "type": "object",
    },
    {
      "errorMessage": {
        "properties": {
          "unstable": "unstable is required",
        },
      },
      "properties": {
        "unstable": {
          "properties": {
            "test": {
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "required": [
        "unstable",
      ],
      "type": "object",
    },
    {
      "errorMessage": {
        "properties": {
          "stable": "stable is required",
          "unstable": "unstable is required",
        },
      },
      "properties": {
        "stable": {
          "properties": {
            "test": {
              "type": "string",
            },
          },
          "type": "object",
        },
        "unstable": {
          "properties": {
            "test": {
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "required": [
        "stable",
        "unstable",
      ],
      "type": "object",
    },
  ],
  "errorMessage": "schema does not apply to stable or unstable",
}
`;
